tests/trx_generator/http_client_async.hpp

```diff
@@ -0,0 +1,166 @@
+   response_callback_t               response_callback_;
+
+ public:
+   // Objects are constructed with a strand to
+   // ensure that handlers do not execute concurrently.
+   explicit session(net::io_context& ioc, const response_callback_t& response_callback)
+       : resolver_(net::make_strand(ioc))
+       , stream_(net::make_strand(ioc))
+       , response_callback_(response_callback) {}
+
+   // Start the asynchronous operation
+   void run(char const* host, char const* port, char const* target, int version, char const* content_type,
+            char const* request_body) {
+      // Set up an HTTP GET request message
+      req_.version(version);
+      req_.method(http::verb::post);
+      req_.target(target);
+      req_.set(http::field::host, host);
+      req_.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+      req_.set(http::field::content_type, content_type);
+      req_.body() = request_body;
+      req_.prepare_payload();
+
+      // Look up the domain name
+      resolver_.async_resolve(
+          host, port, [self = this->shared_from_this()](beast::error_code ec, auto res) { self->on_resolve(ec, res); });
+   }
+
+   void on_resolve(beast::error_code ec, tcp::resolver::results_type results) {
+      if (ec) {
+         response_callback_(ec, {});
+         return fail(ec, "resolve");
+      }
+
+      // Set a timeout on the operation
+      stream_.expires_after(std::chrono::seconds(30));
+
+      // Make the connection on the IP address we get from a lookup
+      stream_.async_connect(results, [self = this->shared_from_this()](beast::error_code ec, auto endpt) {
+         self->on_connect(ec, endpt);
+      });
+   }
+
+   void on_connect(beast::error_code ec, tcp::resolver::results_type::endpoint_type) {
+      if (ec) {
+         response_callback_(ec, {});
+         return fail(ec, "connect");
+      }
+
+      // Set a timeout on the operation
+      stream_.expires_after(std::chrono::seconds(30));
+
+      // Send the HTTP request to the remote host
+      http::async_write(stream_, req_, [self = this->shared_from_this()](beast::error_code ec, auto bytes_transferred) {
+         self->on_write(ec, bytes_transferred);
+      });
+   }
+
+   void on_write(beast::error_code ec, std::size_t bytes_transferred) {
+      boost::ignore_unused(bytes_transferred);
+
+      if (ec) {
+         response_callback_(ec, {});
+         return fail(ec, "write");
+      }
+
+      // Receive the HTTP response
+      http::async_read(stream_, buffer_, res_,
+                       [self = this->shared_from_this()](beast::error_code ec, auto bytes_transferred) {
+                          self->on_read(ec, bytes_transferred);
+                       });
+   }
+
+   void on_read(beast::error_code ec, std::size_t bytes_transferred) {
+      boost::ignore_unused(bytes_transferred);
+
+      if (ec) {
+         response_callback_(ec, {});
+         return fail(ec, "read");
+      }
+
+      // Write the message to promise result
```

Comment:
Comment is wrong, no longer using a promise.

Reply:
Addressed: [9ff417b](https://github.com/AntelopeIO/leap/pull/1052/commits/9ff417bc69e9b87b352ebb9d606c34821547023b)